### ISSUE TITLE
Add opt_size feature for embedded environments.

### ISSUE THREAD

**Issue by cr1901** [jedisct1/rust-ed25519-compact#8]
Tracking issue/continuation from https://github.com/jedisct1/rust-hmac-sha256/pull/2.

Things are not going according to plan tonight. Short version; `ed25519-compact` is very unlikely to fit into the 48kB of easily-accessible flash that I have on my MSP430 part. The microcontroller actually has 128kB of flash (which will _barely_ fit with max size optimizations and the `inline` changes I've made), but I do not know the linker invocation to accomplish this, nor do I know if LLVM meaningfully supports the 20-bit program counter of MSP430. I'll recompile for an ARM micro I have to test speed at some point soon.

In the meantime, please enjoy this image :D:
![image](https://user-images.githubusercontent.com/6418027/154836039-950b4cbc-e9a5-4b6e-aa04-8afe40274c7d.png)

**Comment by cr1901** [comment]
## Benchmarking

I added an `opt_size` feature to my local copy of `ed25519-compact`, and compiled a benchmarking application for RP2040, using the following profiles:

```toml
[profile.dev]
codegen-units = 1
debug = 2
debug-assertions = true
incremental = false
opt-level = 's' # or 3
overflow-checks = true

[profile.release]
codegen-units = 1
debug = 2
debug-assertions = false
incremental = false
lto = 'fat'
opt-level = 's' # or 3
overflow-checks = false
```

`cargo bloat` is the last line of output from `cargo bloat --filter ed25519_compact`, showing the % of the `.text` section that the `ed25519_compact` crate takes as well as the size. It does not necessarily equal the `.text` size column, which is the output from `size`. Here are the results:

|Profile|opt-level|`.text` size|      uS|opt_size?| `cargo bloat`|
|-------|---------|------------|-------|---------|--------------|
|  Debug|        3|      196944| 801755|       No|80.8% 138.5KiB|
|  Debug|        3|      128576| 852571|      Yes| 68.4% 71.2KiB|
|  Debug|        s|      121396| 685952|       No| 64.7% 62.2KiB|
|  Debug|        s|      112180| 732091|      Yes| 60.9% 52.8KiB|
|Release|        3|      149816| 825196|       No|84.5% 118.6KiB|
|Release|        3|       79848| 864777|      Yes| 69.8% 50.2KiB|
|Release|        s|       73420| 659434|       No| 67.3% 44.0KiB|
|Release|        s|       62788| 703593|      Yes| 61.1% 33.5KiB|

Interestingly, the RP2040 is fastest when compiled at `opt_level = s`. The size gains from an `opt_size` feature are more modest here for both `debug` and `release` profiles with `opt_level = s`. 6.5%-7% execution time increase for 8-15% decrease in `.text` size seems to be a reasonable ballpark for the `opt_size` feature.

I don't think it can hurt to include the `opt_size` feature, personally. But a user should still compare sizes rather than unconditionally enabling/disabling the feature.

**Comment by cr1901** [comment]
Closed by #9.
